src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    [Export(typeof(IFileWatcher)), Shared]
+    internal sealed class FileWatcher : IFileWatcher
+    {
+        private ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher> _watchers = ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher>.Empty;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public FileWatcher()
+        {
+        }
+
+        public event ConventionsFileChangedAsyncEventHandler ConventionFileChanged;
+        public event ContextFileMovedAsyncEventHandler ContextFileMoved;
+
+        public void Dispose()
+        {
+            var watchers = Interlocked.Exchange(ref _watchers, null);
+            if (watchers == null)
+            {
+                return;
+            }
+
+            ConventionFileChanged = null;
+            ContextFileMoved = null;
+        }
+
+        public void StartWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
```

Comment:
This highlights some interesting limitations in the available APIs. This would be much easier to read if we could write the following:

```csharp
FileSystemWatcher createdWatcher = null;
var watcher = ImmutableInterlocked.GetOrAdd(
    ref _watchers,
    (fileName, directoryPath),
    key => createdWatcher ?? (createdWatcher = CreateFileSystemWatcher(key.fileName, key.directoryPath)));

watcher.EnableRaisingEvents = true;
if (createdWatcher != watcher)
{
    // If created, the value was not used
    createdWatcher?.Dispose();
}
```

Unfortunately, the deferred detection of `StartWatching` appearing after a call to `Dispose` means we can't use the `GetOrAdd` method in its current form. I'll switch to `lock` for now, and continue to consider a concurrency primitive that provides optimistic concurrency, clarity of implementation and use, and free-threaded single dispose operations.